Web/src/room/RoomPage.tsx

```diff
@@ -0,0 +1,42 @@
+import React, { ReactElement, useEffect, useState } from 'react';
+
+import { Channel } from 'phoenix';
+import RoomContainer, { ErrorCause } from '../components/RoomContainer';
+import Button from '../components/common/Button';
+
+
+const RoomPage: React.FC = () => {
+  const [channel, setChannel] = useState<Channel | null>(null);
+  const [messages, setMessages] = useState<string[]>([]);
+  function renderer(newChannel: Channel | null, error: [ErrorCause, any] | null): ReactElement {
+    if (error != null) {
+      const [cause, _] = error;
+      return <p>Error: {cause}</p>;
+    }
+    if (newChannel == null) {
+      return <p>Establishing connection...</p>;
+    }
+    if (channel == null) setChannel(newChannel);
+    return (
+      <div>
+        <Button onClick={() => newChannel.push('shout', { message: 'Wow!' })}>Shout</Button>
+        <h2>Messages</h2>
+        {messages.map((message: string) => (<p>{message}</p>))}
+      </div>
+    );
+  }
+
+  useEffect(() => {
+    if (channel == null) return;
+    channel.on(
+      'shout',
+      ({ message }: { message: string }) => {
+        setMessages(existingMessages => [...existingMessages, message]);
+      },
+    );
+  }, [channel]);
+
+  return <RoomContainer roomId="lobby" render={renderer} />;
```

Comment:
hmm, why not lift the socket/channel state into a hook instead? I think @afterdusk did that for our assignment 3.

Reply:
What do you mean exactly by hooks? I thought `useState` and `useEffect` are hooks

Reply:
create a `useChannel` or `useRoom` hook that would manage that takes in a room id, and manages the channel state (initiation and connection result) for you. So instead of using this render function pattern, you just call this hook and observe the `channel` and `error` states exposed by it.

https://reactjs.org/docs/hooks-custom.html

Reply:
👍 will implement it this way. Seems much neater